### ISSUE TITLE
[feature request] Support `progressDeadlineSeconds` in CloneSet

### ISSUE THREAD

**Issue by ronaknnathani** [openkruise/kruise#1468]
**What would you like to be added**:
Add support for `progressDeadlineSeconds` in CloneSet. `.spec.progressDeadlineSeconds` is an optional field in [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) that specifies the number of seconds one wants to wait for their Deployment to progress before the system reports back that the Deployment has [failed progressing](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#failed-deployment) - surfaced as a condition with `type: Progressing, status: "False" and reason: ProgressDeadlineExceeded` in the status of the resource.

Reference: https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#progress-deadline-seconds

**Why is this needed**:
In the case of CloneSets, if the rollout is stuck for any reason - image pull errors, insufficient permissions, insufficient capacity, unhealthy app, etc. - currently there is no way to identify that the rollout is stuck by looking at the CloneSet status. With a feature like `progressDeadlineSeconds`, one can specify a time and if the rollout hasn't made any progress for that period, the CloneSet status can be updated to reflect that. The CloneSet controller would still continue to reconcile the requests, however, this status allows continuous delivery pipelines and any other controllers that build on top of CloneSet to know that the rollout failed.

**Comment by ronaknnathani** [comment]
@furykerry do we need a proposal for this or do you think we can send a PR for review?

**Comment by furykerry** [comment]
yes, plz draft a proposal for this feature in https://github.com/openkruise/kruise/blob/master/docs/proposals

**Comment by furykerry** [comment]
@ronaknnathani we're drafting a [proposal](https://github.com/openkruise/kruise/pull/1520) for this feature, however the definition of progressDeadlineSeconds can be different from Deployment because cloneset support partition. Can you give us some feedback about the potential use case of progressDeadlineSeconds when partition is set ?
